zcash_client_sqlite/src/wallet/commitment_tree.rs

```diff
@@ -367,6 +387,46 @@ pub(crate) fn last_shard<H: HashSer>(
     .transpose()
 }
 
+/// Returns an error iff the proposed insertion range
+/// for the tree shards would create a discontinuity
+/// in the database.
+fn check_shard_discontinuity(
+    conn: &rusqlite::Connection,
+    table_prefix: &'static str,
+    proposed_insertion_range: Range<u64>,
+) -> Result<(), Error> {
+    if let Ok((Some(min), Some(max))) = conn
+        .query_row(
+            &format!(
+                "SELECT MIN(shard_index), MAX(shard_index) FROM {}_tree_shards",
+                table_prefix
+            ),
+            [],
+            |row| {
+                let min = row.get::<_, Option<u64>>(0)?;
+                let max = row.get::<_, Option<u64>>(1)?;
+                Ok((min, max))
+            },
+        )
+        .map_err(Error::Query)
+    {
+        if !proposed_insertion_range.contains(&min) && !proposed_insertion_range.contains(&max) {
+            // The proposed insertion range does not overlap with the existing shard indicies.
+            // This means a discontinuity is introduced if the proposed insertion range's start
+            // is not `max + 1` _and_ `min` is not the proposed insertion range's end.
+            if max + 1 != proposed_insertion_range.start && min != proposed_insertion_range.end {
+                return Err(Error::SubtreeDiscontinuity {
+                    attempted_insertion_range: proposed_insertion_range,
+                    min,
+                    max,
```

Comment:
If my nit were acted upon:
```suggestion
                    existing_range: min..(max+1),
```